src/v/storage/disk_log_impl.cc

```diff
@@ -334,15 +334,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
       "[{}] applying 'compaction' log cleanup policy with config: {}",
       config().ntp(),
       cfg);
-    // find first not compacted segment
+
+    // create a logging predicate for offsets..
+    auto offsets_compactible = [&](segment& s) {
```

Comment:
the code is safe. but generally the code base (the exception being tests) avoids capturing the entire context for hygiene reasons.